scio-bigquery/src/main/scala/com/spotify/scio/bigquery/nio/BigQueryIO.scala

```diff
@@ -0,0 +1,414 @@
+        }
+        if (tableDescription != null) {
+          transform = transform.withTableDescription(tableDescription)
+        }
+        data.applyInternal(transform)
+
+        if (writeDisposition == WriteDisposition.WRITE_APPEND) {
+          Future.failed(new NotImplementedError("BigQuery future with append not implemented"))
+        } else {
+          data.context.makeFuture(BigQueryTap(table))
+        }
+    }
+}
+
+object TableRef {
+  sealed trait WriteParam
+  final case class Parameters(
+    schema: TableSchema,
+    writeDisposition: WriteDisposition,
+    createDisposition: CreateDisposition,
+    tableDescription: String) extends WriteParam
+}
+
+/**
+ * Get an SCollection for a BigQuery table.
+ * @group input
+ */
+final case class TableSpec(tableSpec: String) extends ScioIO[TableRow] {
+  type ReadP = Unit
+  type WriteP = TableSpec.WriteParam
+
+  def id: String = tableSpec
+  private lazy val ref = bqio.BigQueryHelpers.parseTableSpec(tableSpec)
+
+  def read(sc: ScioContext, params: ReadP): SCollection[TableRow] =
+    TableRef(ref).read(sc, params)
+
+  def tap(read: ReadP): Tap[TableRow] =
+    TableRef(ref).tap(read)
+
+  def write(data: SCollection[TableRow], params: WriteP): Future[Tap[TableRow]] =
+    params match {
+      case TableSpec.Parameters(schema, writeDisposition, createDisposition, tableDescription) =>
+        val p = TableRef.Parameters(schema, writeDisposition, createDisposition, tableDescription)
+        TableRef(ref).write(data, p)
+    }
+
+}
+
+object TableSpec {
+  sealed trait WriteParam
+  final case class Parameters(
+    schema: TableSchema,
+    writeDisposition: WriteDisposition,
+    createDisposition: CreateDisposition,
+    tableDescription: String) extends WriteParam
+}
+
+/**
+ * Get an IO for a BigQuery TableRow JSON file.
+ * @group input
+ */
+final case class TableRowJsonFile(path: String) extends ScioIO[TableRow] {
+  type ReadP = Unit
+  type WriteP = TableRowJsonFile.Parameters
+
+  def id: String = path
+
+  def read(sc: ScioContext, params: ReadP): SCollection[TableRow] =
+    sc.requireNotClosed {
+      sc.wrap(sc.applyInternal(gio.TextIO.read().from(path))).setName(path)
+        .map(e => ScioUtil.jsonFactory.fromString(e, classOf[TableRow]))
+    }
+
+  def tap(read: ReadP): Tap[TableRow] =
+    TableRowJsonTap(path)
+
+  def write(data: SCollection[TableRow], params: WriteP): Future[Tap[TableRow]] =
+    params match {
+      case TableRowJsonFile.Parameters(numShards, compression) =>
+        data
+          .map(e => ScioUtil.jsonFactory.toString(e))
+          .applyInternal(data.textOut(path, ".json", numShards, compression))
+        data.context.makeFuture(TableRowJsonTap(ScioUtil.addPartSuffix(path)))
+    }
+}
+
+object TableRowJsonFile {
+  sealed trait WriteParam
+  final case class Parameters(
+    numShards: Int = 0,
+    compression: Compression = Compression.UNCOMPRESSED) extends WriteParam
+}
+
+object Typed {
```

Comment:
wooow, I would like to go for a beer with you @jto  just to talk code in this object :).

Reply:
Next time I'm in NYC ;)